storage.go

```diff
@@ -0,0 +1,355 @@
+
+		tpoEndpoint:     newTPR.Endpoint(config.TPO.Namespace) + "/" + config.TPO.Name,
+		tpoListEndpoint: newTPR.Endpoint(config.TPO.Namespace),
+
+		logger: config.Logger,
+		logctx: logctx,
+	}
+
+	// TODO extract init func
+
+	// Create TPR resource.
+	{
+		err := s.tpr.CreateAndWait()
+		if tpr.IsAlreadyExists(err) {
+			s.log("debug", "TPR already exists")
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		} else {
+			s.log("debug", "TPR created")
+		}
+	}
+
+	// Create namespace.
+	{
+		ns := api.Namespace{
+			ObjectMeta: apismeta.ObjectMeta{
+				Name:      config.TPO.Namespace,
+				Namespace: config.TPO.Namespace,
+				// TODO think about labels
+			},
+		}
+		_, err := s.k8sClient.CoreV1().Namespaces().Create(&ns)
+		if errors.IsAlreadyExists(err) {
+			s.log("debug", "namespace "+ns.Name+" already exists")
+		} else if err != nil {
+			return nil, microerror.Maskf(err, "creating namespace %#v", ns)
+		} else {
+			s.log("debug", "namespace "+ns.Name+" created")
+		}
+	}
+
+	// Create TPO.
+	{
+		tpo := customObject{
+			TypeMeta: apismeta.TypeMeta{
+				Kind:       s.tpr.Kind(),
+				APIVersion: s.tpr.APIVersion(),
+			},
+			ObjectMeta: apismeta.ObjectMeta{
+				Name:      config.TPO.Name,
+				Namespace: config.TPO.Namespace,
+				Annotations: map[string]string{
+					"storageDoNotOmitempty": "non-empty",
+				},
+				// TODO think about labels
+			},
+
+			// Data must be not empty so patches do not fail.
+			Data: map[string]string{},
+		}
+		body, err := json.Marshal(&tpo)
+		if err != nil {
+			return nil, microerror.Maskf(err, "marshaling %#v", tpo)
+		}
+		_, err = s.k8sClient.Core().RESTClient().
+			Post().
+			Context(ctx).
+			AbsPath(s.tpoListEndpoint).
+			Body(body).
+			DoRaw()
+		if errors.IsAlreadyExists(err) {
+			s.log("debug", "TPO "+tpo.Name+" already exists")
+		} else if err != nil {
+			return nil, microerror.Maskf(err, "creating TPO %#v", tpo)
+		} else {
+			s.log("debug", "TPO "+tpo.Name+" created")
+		}
+	}
+
+	return s, nil
+}
+
+func (s *Storage) Create(ctx context.Context, key, value string) error {
+	err := s.Put(ctx, key, value)
+	if err != nil {
+		microerror.Mask(err)
+	}
+	return nil
+}
+
+func (s *Storage) Put(ctx context.Context, key, value string) error {
+	var body []byte
+	{
+		v := struct {
+			Data map[string]string `json:"data"`
+		}{
+			Data: map[string]string{
+				key: value,
+			},
+		}
+
+		var err error
+		body, err = json.Marshal(&v)
+		if err != nil {
+			return microerror.Maskf(err, "marshaling %#v", v)
+		}
+	}
+
+	_, err := s.k8sClient.Core().RESTClient().
+		Patch(types.MergePatchType).
+		Context(ctx).
+		AbsPath(s.tpoEndpoint).
+		Body(body).
+		DoRaw()
+	if err != nil {
+		return microerror.Maskf(err, "creating value for key=%s, patch=%s", key, body)
+	}
+
+	return nil
+}
+
+func (s *Storage) Exists(ctx context.Context, key string) (bool, error) {
+	data, err := s.getData(ctx)
+	if err != nil {
+		return false, microerror.Maskf(err, "Exists")
```

Comment:
Is this kind of error handling a new thing? Just asking because it is not consistent and I wonder about the reasoning behind it. For instance it is missing in `Create` and totally different in `Put`. Note that in `List` the string is wrong.